fairscale/optim/oss.py

```diff
@@ -201,6 +203,70 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
 
         return loss
 
+    def clip_grad_norm(self, max_norm: Union[float, int], norm_type: Union[float, int] = 2.0) -> torch.Tensor:
+        """
+        Clip all gradients at this point in time. The norm is computed over all gradients together, as if they were
+        concatenated into a single vector. Gradients are modified in-place.
+
+        Arguments:
+            max_norm (float or int): max norm of the gradients
+            norm_type (float or int): type of the used p-norm. Can be ``'inf'`` for infinity norm.
+
+        Returns:
+            Total norm of the parameters (viewed as a single vector).
+
+        .. note: This is analogous to `torch.nn.utils.clip_grad_norm_` but handles the partitioning and multiple devices per rank
+            under the hood. The default torch util is not applicaple here, because each rank only has a partial view of all the grads
+            in the model, so calling it in the OSS context would lead to different scaling being applied per subset of model parameters
+
+        .. warning: This needs to be called on all ranks, since synchronization primitives will be used
+
+        .. warning: Model paralelism -groups other than world- are not yet supported
+        """
+
+        if self.group != dist.group.WORLD:
+            raise NotImplementedError("Clip norm not yet supported for model parallelism (coming soon!)")
+
+        # Compute the max norm for this shards's worth of gradients
+        max_norm = float(max_norm)
+        norm_type = float(norm_type)
+
+        # Filter out the grad-less params, concatenate params from all devices
+        local_params = itertools.chain(
+            *[
+                list(filter(lambda x: x.grad is not None, device_params[self.rank]))
+                for device_params in self.per_device_params.values()
+            ]
+        )
+
+        # Compute the norm on this grad set
+        local_norm = torch.norm(
+            input=torch.stack([torch.norm(input=p.grad.detach().to(self._device), p=norm_type) for p in local_params]),  # type: ignore
+            p=norm_type,
+        )
+
+        # Sync all the norms from all ranks
+        if norm_type == inf:
+            norms = [torch.zeros_like(local_norm) for k in range(self.world_size)]
+            dist.all_gather(norms, local_norm, group=self.group, async_op=False)
```

Comment:
There is no need to gather all the tenors, since PyTorch distributed supports a `MAX` reduction. As below, you can do
`dist.all_reduce(total_norm, op=ReduceOp.MAX, group=self.group)`

Reply:
makes sense, thanks for the heads up I totally forgot about that